openwhisk/executor.go

```diff
@@ -147,71 +143,105 @@ func consumeStream(streamName string, stream io.Reader, ch chan logging.LogLine)
 	}
 
-	var grp errgroup.Group
-	grp.Go(func() error {
+	// Spawn a goroutine that collects all of the logs and eventually finds the sentinels.
+	// We buffer the logs heavily to ensure that this goroutine can exit as soon as the sentinels
+	// are seen and thus we can rely on a fairly aggressive timeout on it finishing.
+	logs := make(chan logging.LogLine, 128)
+	logCollectionDone := make(chan struct{})
+	abortLogCollection := make(chan struct{})
+	go func() {
+		defer close(logs)
+		defer close(logCollectionDone)
+
 		var sawStdoutSentinel, sawStdErrSentinel bool
-		var errors error
-		for line := range proc.lines {
-			if line.Message == logSentinel {
-				if line.Stream == "stdout" {
-					sawStdoutSentinel = true
-				} else {
-					sawStdErrSentinel = true
-				}
-				if sawStdoutSentinel && sawStdErrSentinel {
-					// We've seen both sentinels. Stop consuming until the next request.
-					return errors
+		for {
+			select {
+			case line := <-proc.lines:
+				if line.Message == logSentinel {
+					if line.Stream == "stdout" {
+						sawStdoutSentinel = true
+					} else {
+						sawStdErrSentinel = true
+					}
+					if sawStdoutSentinel && sawStdErrSentinel {
+						// We've seen both sentinels. Stop consuming until the next request.
+						return
+					}
+					continue
 				}
-				continue
+				logs <- line
+			case <-abortLogCollection:
+				logs <- logging.LogLine{Time: time.Now(), Stream: "stderr", Message: "There was an error while collecting your logs."}
+				return
 			}
+		}
+	}()
 
+	// Spawn a second goroutine to actually write the logs to the destination.
+	errorChan := make(chan error)
+	go func() {
+		defer close(errorChan)
+
+		var errors error
+		for line := range logs {
 			line.ActivationId = metadata.ActivationId
 
 			// TODO: We probably want to do this in parallel. Opting for a simple implementation
 			// first to close the loop. We might also want to check how common multiple loggers
 			// are to justify the added complexity.
-			for _, logger := range proc.remoteLoggers {
+			for _, logger := range proc.loggers {
 				if err := logger.Send(line); err != nil {
 					// We want to continue consuming all of the logs so we don't exit immediately
 					// but surface the error eventually.
 					errors = multierror.Append(errors, err)
 				}
 			}
 		}
-		return errors
-	})
+		errorChan <- errors
+	}()
+
+	if proc.isRemoteLogging {
+		// Write a logging notice into the default streams if the configured logger is not a
+		// remote logger.
+		proc.logout.WriteString(remoteLogNotice)
+		proc.logout.WriteString("\n")
+	}
 
-	proc.logout.WriteString(remoteLogNotice)
-	proc.logout.WriteString("\n")
 	out, err := proc.roundtrip(in)
 
+	// Wait for the sentinels to appear and potentially abort log collection if they don't.
+	select {
+	case <-time.After(noSentinelTimeout):
+		close(abortLogCollection)
+		<-logCollectionDone
+	case <-logCollectionDone:
+	}
```

Comment:
Note: I hate this. We're waiting for logs here and in the invoker now. Feels wrong, but I fear there's no "easy" way out of that 😢